src/scheduler/pbs_sched.c

```diff
@@ -792,45 +506,424 @@ are_we_primary()
+
+	if (poll_context == NULL) {
+		poll_context = tpp_em_init(max_connection);
+		if (poll_context == NULL) {
+			log_errf(-1, __func__, "Failed to init cmd connections context");
+			die(-1);
+		}
+
+		servers = (sched_svrconn **) calloc(1, sizeof(sched_svrconn *) * max_connection + 1);
+		if (servers == NULL) {
+			log_errf(errno, __func__, MEM_ERR_MSG);
+			die(-1);
+		}
+
+		sched_cmds = new_ds_queue();
+		if (sched_cmds == NULL) {
+			log_errf(errno, __func__, MEM_ERR_MSG);
+			die(-1);
+		}
+	}
+
+	connect_server(pbs_default());
+}
+
+/**
+ * @brief reconnect to given server
+ *
+ * @param[in] sconn - pointer server struct to reconnect
+ *
+ * @return void
+ */
+static void
+reconnect_server(sched_svrconn *sconn)
+{
+	close_server(sconn);
+	connect_server(sconn->svrhost);
+	free(sconn->svrhost);
+	free(sconn);
+}
+
+/**
+ * @brief find server structure based on given secondary connection
+ *
+ * @param[in] sock - secondary connection to server
+ *
+ * @return sched_svrconn *
+ * @retval NULL  - failure
+ * @retval !NULL - success
+ */
+static sched_svrconn *
+find_server(int sock)
+{
+	int i = 0;
+	for (; servers[i] != NULL && servers[i]->secondary_sock == sock; i++)
+		return servers[i];
+	return NULL;
+}
+
+/**
+ * @brief read incoming command from given secondary connection
+ *        and add it into sched_cmds array
+ *
+ * @param[in]  sock   - secondary connection to server
+ *
+ * @return int
+ * @retval -2 - failure due to memory optation failed
+ * @retval -1 - failure while reading command
+ * @return  0 - no cmd, server might have closed connection
+ * @return  1 - success, read atleast one command
+ */
+static int
+read_sched_cmd(int sock)
+{
+	int rc = -1;
+	sched_cmd *cmd;
+
+	cmd = malloc(sizeof(sched_cmd));
+	if (cmd == NULL)
+		return -2;
+
+	rc = get_sched_cmd(sock, cmd);
+	if (rc != 1) {
+		free(cmd);
+		return rc;
+	} else {
+		sched_cmd *cmd_prio;
+
+		if (!ds_enqueue(sched_cmds, cmd)) {
+			if (cmd->jid)
+				free(cmd->jid);
+			free(cmd);
+			/* return error if we fail to enqueue first command */
+			return -2;
+		}
+
+		/*
+		 * There is possibility that server has sent
+		 * priority command after first non-priority command,
+		 * while we were in schedule()
+		 *
+		 * so try read it in non-blocking mode, but don't
+		 * return any failure if fails to read, as we have
+		 * successfully enqueued first command
+		 */
+		cmd_prio = malloc(sizeof(sched_cmd));
+		if (cmd_prio == NULL)
+			return 1; /* return 1 as we have enqueued first command */
+		cmd_prio->cmd = -1;
+		cmd_prio->jid = NULL;
+
+		get_sched_cmd_noblk(sock, cmd);
+		/*
+		 * we got priority command,
+		 *
+		 * now check if its restart cycle cmd then ignore it
+		 * as we are not yet in cycle
+		 */
+		if (cmd_prio->cmd == SCH_SCHEDULE_RESTART_CYCLE) {
+			if (cmd_prio->jid)
+				free(cmd_prio->jid);
+			free(cmd_prio);
+			/*
+			 * no return here,
+			 * as we have successfully enqueued first command
+			 */
+		} else if (cmd_prio->cmd != -1) {
```

Comment:
Just a question. Is this else if block required. We can at most get two commands from a single server in the worst case i.e. single normal priority command followed by a super high priority command like SCH_RESTART_SCHED_CYCLE but we never get normal priority command followed by an other normal priority command at a time right ?

Reply:
You are right, but I wrote it so that in future we if ever change that (aka sending non-prio command after prio command or add a new prio command) then we just have to add only one else if for that new command. If you still think we don't need it then I will remove it.

Reply:
Please remove this as it adds an extra confusion to have this code now itself. May be when we have a new super high priority command in future we can add it accordingly.